File: leetcode/Search_a_2D_Matrix.cpp
/* search anti-diagonally */
class Solution {
public:
    bool searchMatrix(vector<vector<int> > &matrix, int target) {
        if(matrix.empty()) return false;
        
        int m = matrix.size();
        int n = matrix[0].size();
        
        int i = 0;
        int j = n-1;
        while(i < m && j >= 0){
            if(matrix[i][j] == target) return true;
            else if(matrix[i][j] < target) i++;
            else j--;
        }
        
        return false;
    }
};

/* binary search based solution */
pair<int, int> search_2D_matrix(vector<vector<int>> &A, int sr, int sc, int er, int ec, int target){
    if(sr > er || sc > ec) return {-1, -1};
    
    int mr = sr + (er-sr)/2;
    int mc = sc + (ec-sc)/2;
    
    if(A[mr][mc] == target) return {mr, mc};
    else if(A[mr][mc] > target) {
        pair<int, int> t = search_2D_matrix(A, sr, sc, er, mc-1, target);
        if(t.first == -1 && t.second == -1) return search_2D_matrix(A, sr, mc, mr-1, ec, target);
        else return t;
    }else{
        pair<int, int> t = search_2D_matrix(A, mr+1, sc, er, mc, target);
        if(t.first == -1 && t.second == -1) return search_2D_matrix(A, sr, mc+1, mr, ec, target);
        else return t;
    }
}

File: leetcode/java/Searcha2DMatrix.java
/*
 * Solution#1 has time complexity of log(m) + log(n)
 * Solution#2 has time complexity of m+n
 */

/* Solution#1 */
// O(lgm + lgn)
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        if(matrix == null || matrix.length == 0) return false;
        int m = matrix.length;
        int n = matrix[0].length;
        
        int l = 0;
        int h = m*n-1;
        while(l<=h) {
            int k = l + (h-l)/2;
            int x = k/n;
            int y = k%n;
            if(matrix[x][y] == target) return true;
            else if(matrix[x][y] < target) l = k+1;
            else h = k-1;
        }
        return false;
    }
}

/* Solution#2 */
// O(m+n) complexity
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        if(matrix == null || matrix.length == 0) return false;
        int m = matrix.length;
        int n = matrix[0].length;
        
        int x = m-1;
        int y = 0;
        while(x >= 0 && y < n) {
            if(matrix[x][y] == target) return true;
            else if(matrix[x][y] > target) x--;
            else y++;
        }
        return false;
    }
}
